commit a1edc9f23fe615b4c8f2323fd9f51a9431dc94fe (refs/changes/49/4749/2)
Author: Pedro Palazon Candel <pedro@joyent.com>
Date:   2018-08-28T18:51:36+02:00 (1 year, 1 month ago)
    
    TRITON-716 Add bhyve to list of allowed brands to do snapshots
    Reviewed by: Marsell Kukuljevic <marsell@joyent.com>
    Reviewed by: Mike Gerdts <mike.gerdts@joyent.com>

diff --git a/lib/endpoints/vms.js b/lib/endpoints/vms.js
index 1f24c5a..7bc2526 100644
--- a/lib/endpoints/vms.js
+++ b/lib/endpoints/vms.js
@@ -24,6 +24,10 @@ var common = require('../common');
 var errors = require('../errors');
 var interceptors = require('../interceptors');
 
+// First Platform Image supporting Bhyve VMs snapshots:
+// FIXME: This is just the latest PI today, needs to be modified
+const MIN_BHYVE_SNAPSHOT_PLATFORM = '20180828T120511Z';
+
 var VALID_VM_ACTIONS = [
     'start',
     'stop',
@@ -1472,7 +1476,7 @@ function createVm(req, res, next) {
  */
 function canSnapshot(vm) {
 
-    if (['joyent', 'joyent-minimal', 'lx'].indexOf(vm.brand) === -1) {
+    if (['joyent', 'joyent-minimal', 'lx', 'bhyve'].indexOf(vm.brand) === -1) {
         return (new errors.BrandNotSupportedError(
             'snapshots are not supported for VMs of brand "' + vm.brand + '"'));
     }
@@ -1495,6 +1499,17 @@ function canSnapshot(vm) {
             + 'supported for docker VMs that use --volumes-from'));
     }
 
+    /*
+     * Not all the Platform Images will support bhyve snapshots. Let's fail
+     * early instead of create a Job which will fail anyway
+     */
+    if (vm.brand === 'bhyve' &&
+        vm.platform_buildstamp < MIN_BHYVE_SNAPSHOT_PLATFORM) {
+        return (new errors.BrandNotSupportedError('snapshots are not '
+            + 'supported for bhyve VMs unless Platform Image is "' +
+            MIN_BHYVE_SNAPSHOT_PLATFORM + '" or newest'));
+    }
+
     return true;
 }
 
